### ISSUE TITLE
Improve AI

### ISSUE THREAD

**Issue by dnk777** [Qfusion/qfusion#307]
I have a plan to significantly improve AI. About 20% or planned changes are made, and progress will depend heavily of my load at work, but current changes are already worth to see. My code is in very early stage, and contains well-known bugs (a memory corruption source of what I did not spot yet), but I cant wait to share progress. The code is in development in my fork.

Done:
- Bots never did bunnyhopping at least in 1.5+, fixed.
- Old bot area checks were stupid and caused bots getting blocked in walls (`AI_CheckMove` just checked a trace from eyes to forward ground, and detected walls as a ground).
- Old bots have dumb target and weapon selection, they used stick to a last attacker. My code uses quite sophisiticated `EnemyPool`, that tracks seen enemies, last targets and attackers, suggests targets and weapons using complex heuristics based on distance, health status, powerups, direction, gravity and environment, and suggests advance/retreat/inhibit fight tactics. Bots now do weapon combos to finish enemies. Bots now try to spam last seen enemy positions if there is no enemy in view.
- Old bots never evade projectiles due to a bug. Bots now may evade rockets, gunblade blasts, grenades, plasma streams and laser beams. Plasma projectiles are clusterized in rays based on direction and proximity. Bots evade and dash heavily in fight.

Short-term goals:
- Fix my bugs :)
- Introduce more human-like hitscan aiming by calculating speed of change of dot products between look dir and dir to enemy.
- Indroduce "true" strafe-jumping as described in "Strafing theory" by injx, its a 5-minute task itself but depends of fixing bot mouse movement (bots roll mouse like a drunk each second atm).

**Comment by adem4ik** [comment]
- Bots never did bunnyhopping at least in 1.5+, fixed.

That isn't really a bug, but the feature. Vic decided to disable bot's bunnyhopping for the gametypes with items in order to prevent that issue - https://github.com/Warsow/qfusion/issues/134 - and some more

**Comment by dnk777** [comment]
Idk is it really caused by bunny, with a planned improved movement it is likely to be mitigated anyway. I suggest you try my fork anyway, even if it is in very early stage, I'm going to make bots a real beasts.

**Comment by viciious** [comment]
Not to sound discouraging but the Vec3/Vec3Like/Vec3Ref stuff is atrocious and has zero chances to ever get merged.

**Comment by dnk777** [comment]
Well, please explain how may I fix that

**Comment by viciious** [comment]
- We already have C functions for vectors, use them instead.
- I'm not getting the Ref an Like parts either. Vector is a simple scalar class, there's absolutely no need to have two additional layers of classes.
- Vector classes if implemented, should be templated anyway.

**Comment by dnk777** [comment]
Vec3 owns its own plain old array.
Vec3Ref wraps a pointer to a plain old array.
Vec3Like is a supertype for Vec3's and Vec3Ref's that contains common vector ops.
Vec3Like have a pointer to array, but does not know whether it owns the array or not.
If vector op needs its own buffer, it always return Vec3.
Vec3Ref have been introduced to make a distinction from Vec3Like that may or not own its array.
Vec3Ref is quite similar, to, for example, QLatin1String from Qt and designed to work C++ way with mutable references to old code vectors without touching old code.
I thought about adding glm library in a project, but decided that introduction of these wrappers is less invasive and used 'em only in /ai internals

**Comment by dnk777** [comment]
Its my development sandbox, its not going to be merged right now, I did not make a pull request. We'll discuss my results later when I do it completely

**Comment by viciious** [comment]
Alright.
It'd be nice if you could record some botmatches to demonstrate the AI improvements.

**Comment by dnk777** [comment]
Its very incomplete and bugged atm, but playing it if it does not crash is funny. Planned completion time is the June 1th. I have added quite lots of code having some spare time these days, will be under heavy work-load next months.

**Comment by dnk777** [comment]
Upd: pushed new code, now bots try to aim like humans. 

Old code just added random noize in target position in attacking frames, some sinusoidal noize when firing LG. Bot with a lasergun was a complete joke. 

Let me explain changes.

Human hits 100% when target on screen does not move. Real accuracy on moving target is limited by brain and muscular latency. Bots have a source of latency too, the limited yaw/pitch speed used in `ChangeAngle`. Now bot tries to always keep its lookdir on enemy. We still add some random noize, but not only in attacking frames.

It needs tweaking as everything else, but now bot aim both much more deadly and human-like.
https://github.com/dnk777/qfusion/commit/7b760cff1d11e0f5eda13e66d648aba9c8b3ccdb

**Comment by dnk777** [comment]
There are not critical bugs (can't trigger AddressSanitizer anymore) and I strongly suggest everybody try to play my fork.

Aside from bug fixes a new feature has been added. Now bots try to detect escaping enemies, fast switch weapon if it is possible and hit escaping enemy hard in a single shot. https://github.com/dnk777/qfusion/commit/f537366186e1d8358873bc26b8accf26167053cd

There are not planned major changes in target/weapon selection/combat code anymore aside from minor bugfixes and values/condition tweaking.

Next goal is to improve bot movement, but it is deferred at least at a single month.

March TODO's:
- Track changes in old `edict_t` fields that may be set by scripts, I am not going to change the script interface.
- Add some kind of `pendingLookDir`, `pendingLookAtPoint`, very useful for turning in combat, rj, wall-rj (in future).
- Extrapolate last seen enemy positions and set extrapolated one as `pendingLookAtPoint`, bots often loose enemy out of view atm

**Comment by solidfake** [comment]
Hmm, despite the obvious issue of bots sometimes getting "stuck" into walls and loosing their will to hunt enemies, my main issues with bots are:
-the constant flickshooting, bots don't aim where they see that the enemy player is, but where they know where he is (minus accuracy depending on skill), it looks extremely awkward when spectating bots as they constantly lock onto every player in sight even if he is 180° behind them. But that maybe needs a bigger rework, as the only skilllevel bots really have now is their accuracy.
-They still have extreme issues with everything that involves a walljump to get at, which makes most maps quite bland when you know they will never contest that RA or reach that platform

But I have no say in what gets approved anyway, just a heads up if you still not feel demotivated enough after fixing your mentioned stuff.

**Comment by viciious** [comment]
I'll check out this branch whenever my time permits.

**Comment by viciious** [comment]
I can't say I'm totally happy with STL creeping in as well.

**Comment by dnk777** [comment]
> the constant flickshooting, bots don't aim where they see that the enemy player is, but where they know where he is (minus accuracy depending on skill), it looks extremely awkward when spectating bots as they constantly lock onto every player in sight even if he is 180° behind them.

Its just kept old bots behaviour, they do not check even a dot product between look dir and enemy dir, they just use `inPVS`.
https://github.com/Warsow/qfusion/blob/master/source/game/ai/ai_class_dmbot.cpp#L961
https://github.com/Warsow/qfusion/blob/master/source/game/g_utils.cpp#L1728
 It will be reworked to lock on only really visible ( = in field of view) enemies, back enemies will be detected when a bot may hear their shots/jumps or see projectiles coming from behind. In this case, I will set a guessed enemy position as a `pendingLookAtPoint` (to be introduced).

> They still have extreme issues with everything that involves a walljump to get at, which makes most maps quite bland when you know they will never contest that RA or reach that platform

I did not touch movement significantly yet,  and, yes, its just wrong it its basics atm. While I have allowed bots to bunnyhop, it its not very useful atm, since bots hit walls all the time due to following reasons:
- Bots check obstacles only on contact.
- Navigation path provides only a coarse direction of movement. 

Environment will be tested and for a current path segment (-1 nav node back, +1 nav node forward) a detailed, obstacles-aware movement path will be built.

> I can't say I'm totally happy with STL creeping in as well.

STL is to be removed from the ai subdirectory aside from things that do not allocate anything (`std::max`, `std::find`, etc). Atm the only container used in code is `std::deque`, I wanted something reliable to start with, it is to be replaced by a fixed-capacity circular buffer.

**Comment by dnk777** [comment]
UPD: Buggy attempts to introduce advanced bot movement were reverted.
To give bots a chance anyway, I have enabled forward bunny for bot clients.
Now bots are OK for FFA (not DM, item pickup code remains untouched and it is weak).
There are known bug that often bots start to ignore weapon/target choice timeout, but it does not ruin gameplay.

**Comment by adem4ik** [comment]
Will be there any performance regressions after applying your improvements, @dnk777 ?

**Comment by dnk777** [comment]
> Will be there any performance regressions after applying your improvements

For servers, definitely yes, but there are huge performance reserves anyway and the code is not optimized yet.
Bot code runs on server and does not affect clients (if it does not hang a server :)

**Comment by dnk777** [comment]
There are following issues that prevent implementing really good AI (mainly movement):
- Bot doesn't have a knowledge about a map environment, it is like a blind man that tries to touch environment with a walking stick. This affects not only movement but target/weapon/tactics selection too.
- Current paths that consist of nodes are too restrictive, we have to follow the path nodes and try to complete them. Thats why I have to remove strafe-jumping code I have added once. You may see that in fight bots move much more vivid and properly, since they are not restricted by nodes and trace environment heavily. I want to know that a bot is in a some box volume with a floor and walls, and be able to use any arbitrary movement path as long it is inside the volume and goes from its entrance to an exit.

Thus I wont touch movement code anymore, aside from bugfixes in current code while there are no rich navigation support in a map data. I think about porting AAS code from Q3A, it would take not less than 2 month given my limited amount of time I may spend on modification and poor engine knowledge.
AAS runtime is to be ported first, Q3 maps with precompiled AAS may be used for testing.

**Comment by viciious** [comment]
The code for q3 bots navigation is monstrous. Porting it to qfusion would be a no small feat.

**Comment by dnk777** [comment]
AAS code builds well together with the qfusion source, the only thing needed is to fix some includes/defs. I didn't try to integrate or run it yet, just compiled it.

**Comment by viciious** [comment]
Gonna post this link here for reference purposes: http://fd.fabiensanglard.net/quake3/The-Quake-III-Arena-Bot.pdf

**Comment by dnk777** [comment]
I have read this paper already. And yes, the ioquake3 AAS code that I use looks monstrous and scary, but I used for overview https://github.com/id-Software/DOOM-3-BFG/blob/master/neo/aas/AASFile.h, they refactored it over the years.

**Comment by dnk777** [comment]
AAS files are loaded well (I've tried all cpma mappack), atm I just initialize/shut down AAS code and load a map. I have a question, how to implement AAS debug drawing https://github.com/dnk777/qfusion/blob/a5a54ffb5fac083e404be6de45484ec4d197326e/source/game/ai/aas.cpp#L217 (how to check whether I have a client attached and draw via the client, spawning server events is not an option since its very limited)? I need debug visualisation asap, I want to see what the AAS thing is.

**Comment by viciious** [comment]
The best approximation would be to spawn an ET_LASER entity in AAS_QF_DebugLineCreate, delete it in AAS_QF_DebugLineDelete and assign draw parameters AAS_QF_DebugLineShow

**Comment by dnk777** [comment]
Quick report: AAS works well, the hardest part is to figure out how to get a chain of AAS areas connected by reachabilities that leads to target, Q3 interface is quite confusing. Introduction of AAS is the most radical change I made, all nav-node-related code has been removed, and it touches everything. Atm the code is in my local repo, I continue to re-implement old code using AAS now.

http://webmshare.com/K7O7q

On the video combat is disabled, and there are known bugs (bot touches items, and when he can't pick them, he gets assigned the item as a goal again since the last goal item is the closest one, pickup code will be improved and on-pickup-actions will be introduced (e.g. damage himself to pick armor)). I had to shoot the bot to make him pick up health, and kill him since he can't pickup RL ammo anymore. Debug drawing is my custom, gold things are `goalTargetPoint`'s in `goalAasArea`, purple ones are analogs of old `self->ai->move_vector`, blue ones of various length show AAS reachability vectors.

**Comment by viciious** [comment]
The navigation towards goals is quite odd, the bot is noticibly circling around them before finally picking them up.
I recall a similar problem I observed with bot movement: they tried to move towards the goal while rotating the view and ended up spiraling around just like in your video.

**Comment by dnk777** [comment]
Yes, its like a movement in a planet gravity field, if bot has a "tangential" velocity, he keeps walking around the target, its a control theory problem. I have already added a workaround (otherwise bot walked around the target forever), its just not tweaked enough for allowed bunny speed. The workaround is checking dot product between actual move dir and a dir from self to target if we are close to target, if its less than a threshold, do one of the following things:
- Roll view harder
- Determine on which side of move dir line the target is. Then release +forward and press strafe to the target side, bot will turn using aircontrol.

**Comment by dnk777** [comment]
Upd: Basic running movement has been implemented (I did not touch juppads, platforms, etc.). The code is still in my local repo. It it very dodgy atm, since bots tends to stick to nearest AAS reachability, as they used to do it with old nodes. However, it is still a great improvement and a great simplification, we do not need to check whether we have touched a node or not, we may just check AAS area number of current origin point (which is quite cheap, a single BSP lookup).

http://webmshare.com/nrgyr

I plan to smooth actual bot movement path setting current move dir as a weighted sum of dirs to current and some next reachabilities with decay, then add «true» strafejumping (atm it may be implemented but will not be very useful, bots will bump into walls permanently).

I have questions: 
- How to implement tracking of allocated/deallocated memory for AAS subsystem. I have to provide real available for AAS free memory amount instead of this stub https://github.com/dnk777/qfusion/blob/aas-port/source/game/ai/aas.cpp#L186, otherwise AAS subsystem never clear cached routes. The problem is how to retrieve size of memory chunk freed in https://github.com/dnk777/qfusion/blob/aas-port/source/game/ai/aas.cpp#L181. Is exposing `memheader_s` defined in https://github.com/Warsow/qfusion/blob/master/source/qcommon/mem.c#L33 to `qcommon.h` an acceptable thing?
- How to implement arbitrary debug AAS drawing on a listen server. Using broadcasted events is not an option, since they cause edicts overflow, and I want to be able to draw polys too. The problem is how to add debug drawing to a server code (that may be exported by a game module). May I add something like `DrawDebugLine`, `DrawDebugPoly` like, for example, `SCR_DrawLoadingPlague` in https://github.com/Warsow/qfusion/blob/master/source/qcommon/qcommon.h#L1029` and use stubs for dedicated servers?

**Comment by viciious** [comment]
- No, you should not include qcommon headers, which are private to the engine core, from game modules. Just get rid of *AvailableMemory and *HunkAlloc calls altogether, this is ancient stuff. 
  You could also wrap the allocated chunk into helper struct:
  typedef {
  int chunk_size;
  void *chunk;
  } helper_foo;
  and return pointer to chunk. Then, restore the pointer to helper_foo \* in free.
- The short answer, yes, that would be the best approach. You'll need to proxy-pass your DrawDebug\* calls down to cgame, to get your debug stuff rendered properly.

**Comment by dnk777** [comment]
`HunkAlloc` is used by AAS subsystem to load AAS data, i backed it by `G_LevelMalloc`.
Having valid `AvailableMemory` is mandatory, otherwise AAS routing code will not clean cache and will allocate chunks forever:
https://github.com/ioquake/ioq3/blob/master/code/botlib/be_aas_route.c#L1624

**Comment by viciious** [comment]
Well, you can always change those few lines to work differently.. Otherwise, create a separate memory pool for AAS and track its memory usage like I outlined above.

**Comment by viciious** [comment]
Here's another solution: 
add the following function to g_utils.c:
`static int G_Z_AvailableMemory( const memzone_t *zone )
{
return zone->size - zone->used;
}`
and use it for G_LevelAvailableMemory and then for AAS_MemoryAvailable.

Use a separate memzone akin to levelpool for AAS hunk allocations. Free it along with the levelpool.

**Comment by dnk777** [comment]
Recent movement video, looks promising, even GREAT considering how broken bot movement used to be: http://webmshare.com/VN95P. I have not implemented yet neither climbing/jumppad/platform/RJ movement, nor even walljumping, its just generic running. I'm going to push first public version of AAS port in 1-2 weeks.

**Comment by adem4ik** [comment]
would be nice to see bot's POV video

**Comment by viciious** [comment]
Nice!

**Comment by dnk777** [comment]
Upd: I decided to push the AAS code in public access though it is incomplete and poor yet. 
https://github.com/dnk777/qfusion/tree/master
Combat is still disabled and only minimal movement styles set (running/bunnying/jumppad using) has been implemented. And this change was very invasive, I had to reimplement half of all AI code and break script compatibility.

Known bugs: sometimes bots stand still blocked and blocking recovery code is not triggered until suicide, source of the bug has not been spotted yet.

Known issues (not a bugs, need to be tweaked): 
- Bots may fail in gaps (i try to shortcut movement path, and there are not cheap ways to check for ground/failure), path straightening/shortcut code is very basic.
- Bots may spin weird, it is an attempt to implement 180-degree turns with dash (when you release all keys in air, then turn mouse very quickly to a target and hit dash on landing), it works but looks robotic.
- Bots use fake acceleration, it is probably bad thing for netcode prediction, but since bots are in development and are tested only on localhost it is not a big issue. Implementing a true strafejumping is not a big deal though.
- The code still leaks (AAS routing cache is never cleaned up), I did not implemented proper AAS allocations handling yet.
- Combat compatibility has not been tested yet.

http://webmshare.com/R5Dq7

Overall the current bot movement is far from being complete, and I do not consider it not good, not even satisfactory. Frankly the code is a one big hack, since AAS provides good support for bots … walking on ground with 320 ups, i just try to fly/bunny over the map and check for turns/obstacles in runtime. I am surprised it even works overall. AAS however is still better than old nav nodes since AAS works with (convex) volumes, not points and greatly simplifies reachability checks.

**Comment by dnk777** [comment]
Let me describe upcoming changes... I think about 60% of work is done. Release is still deferred to June. Aside from debugging and polishing current code, i plan to introduce multi-level AI brain.
- AiBaseGametypeBrain. This entity should exist as a single instance and think once per frame. It should be used for knowledge exchange among all AI entities present in game regardless of their team and for interaction with gametype, e.g. for cancellation of goals reached by a single bot for all AIs, searching for noobs by rankings to hunt, etc.
- AiBaseTeamBrain. A team is a group of bots (and humans) that share score and strategic goals. It should group bots in squads and determine strategic goals (e.g. schedule powerup pickup, distribute armor pickup between members of a team, etc).
- AiBaseSquadBrain. A squad is a group of bots (and humans) that roam and fight together. Squad brain should suggest tactics: relative and absolute position of bots, prioritize enemies, etc to implement behaviors: «assist carrier», «assist item pickup», «prevent capture», «prevent item pickup», «drop and pickup weapon», etc.  
- AiBaseBrain - already exists with its subtype, the BotBrain.

Of course it should be made scriptable, but some basic skeletons for gametype («basic ffa gt brain», «basic TDM-like gt brain», «basic round-based gt brain»), team and squad brain will be provided in AI code.  

It is easier to implement that it sounds (it is a pure programming logic, no need to dive in physics simulation, dated Q3 code, fight bot blindness in 3D world, etc).

And of course I am acknowledged about current performance… issues, but won't fix them by distributing workload between frames until the final control flow will be known (obviously team/squad members should think in same or consequent frames).

**Comment by Mark90** [comment]
Sounds awesome. Looking forward to seeing the result both in code and ingame :)

**Comment by viciious** [comment]
@dnk777 Any status updates?

**Comment by dnk777** [comment]
https://github.com/dnk777/qfusion/commits/master.

I fixed some bugs, mainly performance issues by introduction of AiBaseTeamBrain.
That class is a stub atm, it only manages think frames distribution between members of a team (which is not its direct purpose but it is required for planned changes), but does not provide any teamplay-related logic. There is a known bug, gametype change is not tracked properly (I do not know the engine control flow well). How may I contact dev team to ask dumb questions instead of spamming in this thread?

Also, this https://github.com/dnk777/qfusion/commit/6537cd5e3ea73169843edc63b734124bfaf1a8db commit fixed critical performance drops (N^2 trace count explosion) 

Bots are playable in FFA (not DM). 

I think I should defer complex teamplay logic implementation, bots are too weak individually, they can't even pursuit enemies atm. In worst case I think individual bot logic will be improved and polished in these 2 months.

**Comment by viciious** [comment]
> How may I contact dev team to ask dumb questions instead of spamming in this thread?

I think you could instead open new issue(s) for your questions. I'd prefer to keep our communication public, as a sort of documentation process.

**Comment by viciious** [comment]
I've also started working on qfusion/warsow-compatible version of bspc, should be completed sometime this week.

**Comment by viciious** [comment]
https://github.com/dnk777/qfusion/blob/master/source/game/ai/aas/be_aas_reach.c#L2468

```
        ladderface1vertical = abs(DotProduct(plane1->normal, up)) < 0.1;

        ladderface2vertical = abs(DotProduct(plane2->normal, up)) < 0.1;

        //there's only reachability between vertical ladder faces

        if (!ladderface1vertical && !ladderface2vertical) return qfalse;

        //if both vertical ladder faces

        if (ladderface1vertical && ladderface2vertical

                    //and the ladder faces do not make a sharp corner

                    && DotProduct(plane1->normal, plane2->normal) > 0.7

                    //and the shared edge is not too vertical

                    && abs(DotProduct(sharededgevec, up)) < 0.7)
```

This whole block confuses abs with fabs multiple times

**Comment by dnk777** [comment]
> This whole block confuses abs with fabs multiple times

I think it worth sending a PR to upstream/ioq3, isn't it? And it is not an AAS runtime code, so it does not affect bot runtime behaviour aside from providing some buggy AAS files.

Some notes on AAS:

There are travel flags that correspond to AAS reachabilities: https://github.com/Qfusion/bspc/blob/master/deps/botlib/be_aas.h#L37 I have never see any Q3 maps (including CPMA ones) that contain strafejump, rampjump or doublejump reachabilities. AAS compiler code that determines reachabilies is located in https://github.com/Qfusion/bspc/blob/master/deps/botlib/be_aas_reach.c Strafejump reachability may be added by modifying https://github.com/Qfusion/bspc/blob/master/deps/botlib/be_aas_reach.c#L2103 for increased gap distance. Also, for very large gaps that may be shortcut using horizontal weaponjump (or just very high speed) a new reachability (horizontal weaponjump) may be introduced. Also, you may try to modify jump reachabilities and introduce new walljump reachabilities when target area is higher than current area and a side wall exist. AAS compiler and runtime code are not dependent, so only an introduction of new travel flags may require runtime code patching as well.

The original code searches for reachabilities to a target area only if target area contains an important item, and item names are hardcoded. https://github.com/Qfusion/bspc/blob/master/deps/botlib/be_aas_reach.c#L3938. I think that GB-jump reachability may be introduced, and a teleport entrance is a target that worth weaponjumping too.

AAS files may be optimized(stripped): almost all area faces and edges (`aas_face_t`, `aas_edge_t`) are removed, and only coarse AAS area infos (center, AA-bounds and flags) and AAS reachabilities are left. I strongly suggest to avoid stripping for newly compiled maps, AAS geometry data is/would be extremely useful and may replace many runtime traces in bot code which are expensive. For stripped maps (if someone loads Q3 stripped map) we have to fallback to traces, and due to very limited traces count and coarse nature of traces it may cause poor bot behaviour.

**Comment by viciious** [comment]
> I think it worth sending a PR to upstream/ioq3, isn't it?

I will submit the fix to upstream bspc.

Here's FBSP-aware build of bspc which can be used to produce AAS files for Warsow maps in the following fasion: 

> ./bspc -bsp2aas basewsw/map_wbomb6.pk3/maps/wbomb6.bsp -forcesidesvisible

https://github.com/Qfusion/bspc/tree/qfbsp

**Comment by viciious** [comment]
Finally tried this thing.. 
I couldn't quite get what the bots were up to all the time with they random movement and missing item pickups. They also seem to enjoy aiming straight up alot. 
The aas/botlib code will need to be moved to its own library so it doesn't pollute the actual game code.

**Comment by viciious** [comment]
@dnk777 May I suggest you to resume your development off a different branch? I've managed to rebase your work off qfusion master: https://github.com/Qfusion/qfusion/tree/dnk777_rebased

**Comment by dnk777** [comment]
> I've managed to rebase your work off qfusion master

Ok thanks, got it.

> The aas/botlib code will need to be moved to its own library so it doesn't pollute the actual game code.

The AAS code came with its own exports, but I found these exports not useful, since I need access to AAS guts. While I might modify exports, I thought that I should avoid extra work maintaining exports that are changed frequently. I will separate AAS code and restore exports when the final AAS interface required for AI code is known.

> Bots have weird behavior

I did not claim that the work is either finished or even is satisfactory. I told planned release date many times. I have just started a work on the bot logic and not on mandatory but auxillary code.

**Comment by viciious** [comment]
Ye-ye, I know. I'm not criticizing the WIP, just sharing my impressions of this stage of development, as well as pointing out the flaws which otherwise might get lost.
I will also try the master branch again later today to see if it's any better than the aas-port branch.

**Comment by dnk777** [comment]
> I will also try the master branch again later today to see if it's any better than the aas-port branch.

This means that you played working (that does not crash) but unplayable branch.
Master is two weeks of bugfixes ahead of it.

**Comment by viciious** [comment]
Just tried the master branch and failed to notice any difference: bots still jump around the corners, aimlessly turning their heads into random directions.
At first I thought it was my buggy bspc producing bad AAS files for warsow maps but stock q3a maps exhibit the same bot behavior.

**Comment by dnk777** [comment]
The updated branch is there: 
https://github.com/dnk777/qfusion/tree/dnk777_rebased

I do not know what you mean complaining about bot behavior. Its of course weird but is a good point to start improve it for me. http://webmshare.com/jjD1g

Most of AAS files produced by BSPC can't be loaded. I tried generating AAS for wdm1, wdm7, and it fails with something like this:

```
Gametype 'Free for all initialized'
--- Botlib initialization ---
----- Map loading -----
trying to load maps/wdm1.aas
**********
ERROR: Game Error:  G_Z_Malloc: failed on allocation of 230193 bytes
**********
==== G_Shutdown ====
```

wdm9 though works fine. I did not try loading any Q3 map yet.

**Comment by viciious** [comment]
I'm not complaining, just pointing out. Try speccing a lone bot to see him suffer through the map.

> I tried generating AAS for wdm1, wdm7, and it fails with something like this:

In g_local.h, change G_LEVELPOOL_BASE_SIZE to 15MB

**Comment by dnk777** [comment]
To have not-looking-weird bot behavior I have to implement seamless transition between combat and roaming movement. Roaming movement is planned via AAS subsystem. Combat movement is random and only affected by "attack"/"retreat" flags in `CombatTask`, it is not aware neither of map topology, not of goal entity origin. Bot should not lose goal in combat movement, but to implement it I have to make much more sophisticated goal planning with cancellation of goals (when bot is weak) and the enemy blocks it. 

That is a recent video of "camping a spot" movement which is not used atm (I forced it for debugging purposes in private build), but is very useful for waiting for goal pickup (item timing is a part of goal subsystem improvement and will be added soon). http://webmshare.com/jjbLO The code that produces this kind of movement is here: https://github.com/dnk777/qfusion/commit/b28f8dc14d68e679ed25f881c0296ebee0712760 The video shows "camping a spot" movement with different parameters (alertness and intended look-at-point, specified or not). Imagine a bot doing this waiting for armor pickup/bomb plant.

**Comment by viciious** [comment]
Imo improving combat or camping AI should be of lesser concern at this point... http://webmshare.com/DA5XG

**Comment by dnk777** [comment]
Half of these bugs I guess were fixed last day, and roaming movement will be improved anyway for the purpose of combat and roaming movement fusion. And that "camping a spot" movement is very useful not only for combat/goal actions, but for roaming movement too (a bot will move this way on an elevator platform).

**Comment by dnk777** [comment]
How do you look at a bot cheating with movement, if it is done carefully and not very-robotic looking? I mean a bot that has an ability to set both magnitude and direction of `self->s.velocity` while in air (`self->groundentity == NULL`, ground/combat movement is to be kept as is)? The problem is with trying to control a bot that tries to follow usually curved AAS reachability chains with an ordinary movement keys and thus fully affected by the player physics with inertia. You saw that even making a bot to pick an item is quite complicated due to control inertia. I know that the praised CPMA/Spiterbots that are able to imitate a high-skilled in movement player cheat with movement. Breaking client-side player prediction is the downside of this approach.

**Comment by viciious** [comment]
That should be the last resort.. As of now, I can't see how the problem lies with the aircontrol inertia. The problem as I see it, is getting bots to travel from point A to point B. Enabling bunnyhopping movement prior to getting the very basic things to work was a mistake.
Bots have a hard time picking up items even if they are in the same AAS area, even if it would require walking in a straight line, which they attempt and just walk past the item very often.

**Comment by dnk777** [comment]
> Bots have a hard time picking up items even if they are in the same AAS area,

Was fixed in https://github.com/dnk777/qfusion/commit/c4ba52439983ee9f64dd05e7ca8d8054705fb197.

The problem is in areas size, they are usually too small and are changed too often during fast movement. An area is a simple convex volume determined automatically. Where a human sees a large room, e.g in a quad-to-tele area on wdm9, bot has lots of small convex areas connected by reachabilities that are determined by bspc too. I'd rather prefer navigation volumes (not nodes, volumes are much, much superior due to simplicity of position/reachability checks!) that are placed manually by a mapper but writing an editor for it is not a rewarding thing. The situation is worse on the maps with more complex collision geometry, they behave not so bad on wdm9. I either try to follow AAS reachabilities chain, which is rarely straight and for small areas require _lots of control_ in movement or try to straighten the path through areas as long as a straightened path follows these areas (this code is now turned off due to amount of movement mistakes it introduces; there is no a cheap way to check whether straightening a path/shortcutting is legal).

I'll try it in there weeks... good players are able to do some «supernatural» moves anyway. I believe a bot that moves fast and precise but, maybe, a bit unnatural is better that a bot that cannot bunny. Is misleading the netcode prediction the only thing that may have a negative impact? And netcode prediction impact is to be proven yet, considering relatively high server rates the game runs on.

**Comment by viciious** [comment]
> Was fixed in dnk777@c4ba524.

No it wasn't. On, say, wdm1, bots miss the quad and ra in like 70% of cases, they just walk or jump past by.
Overall, their navigation around wdm1 is so bad it actually hurts.

The way they manage to miss the damn quad or the ultra on wdm9 (sometimes two or three times in succession) is equally amusing.

The way they manage to dodge the damn quad on wdm9 multiple times in a row is also pretty amusing.

> And netcode prediction impact is to be proven yet, considering relatively high server rates the game runs on.

You could enable the pmove flag that disables prediction for bots (grep for PMF_NO_PREDICTION). That would fix it for specs, at least.

But trust me, you would _not_ want to play with movement prediction turned off. Try cg_predict 0 and connect to any server.

**Comment by dnk777** [comment]
Does the game use enemy ucmd in prediction of enemy movement for a client? If it just (linearly) extrapolates its entity params prediction errors should be negligible considering high server rate. I do not mean turning off movement prediction for an active POV itself!

**Comment by viciious** [comment]
Ah, ok, it should be look/work fine then.
You still going to need PMF_NO_PREDICTION for bot clients though :)

**Comment by dnk777** [comment]
On the video http://webmshare.com/K7oBV these cyan lines correspond to AAS reachabilities (`aas_reachability_t::start`, `aas_reachability_t::end`) from some area to another one. 

```
for (int i = 0; i < aasworld.reachabilitysize; ++i) {
   auto &reach = aasworld.reachabilities[i];
   if (random() < 0.01) // events flood protection
      // spawn a laser event from reach.start to reach.end
}
```

They are shown as server-side events and due to limited events count they blink. You can see that each stair step (and even a part of a step) is treated as a separated area. Writing a code that tries to change `intendedLookVec` by either following these reachabilities or area centers, or short-cutting and straigtening the path is complicated. The March version that looked satisfactory was developed on low-poly Q3 maps. I should develop a code that tries to short-cut this mess instead of following each reachability.

**Comment by viciious** [comment]
Try different bspc settings maybe?

**Comment by dnk777** [comment]
I should redo `intendedLookVec` calculation anyway. Only area centers/faces (the problem with faces is their absence in optimized/stripped AAS files, loading these Q3 maps should not be supported) that are volumetric should be used instead of reachabilities that are a single line segment that consist of two points mentioned above and thus lead to curved movement path and return us to a node/point reachability hell.

**Comment by dnk777** [comment]
I have fixed most annoying movement bugs, but bot movement still is not more than mediocre.
Bots should perform better on simplistic, low-collision-poly maps.

Look at the video I have posted days ago, maps are full of junk areas. 
AAS compiler treats each horizontal surface as a navigable area, you may notice on the video that even top of the railings may be involved in pathfinding.
These small areas can't be easily removed, because it may break map topological connectivity.
Bot navigation is a mapper concern too, from what I read in bspc readme a mapper may prevent navigation in some areas by marking parts of a map as «do not enter».

I will try to tweak preferred and allowed travel flags used in pathfinding to cut off these kinds of reachabilities that often go along with junk areas.

**Comment by dnk777** [comment]
I have implemented seamless combat/roaming movement and goal rejection/cancelation
when a bot should retreat and path to a goal is blocked by enemy. These improvements are shown on the video.  

http://webmshare.com/bLOwb

 Those pink lines on video are drawn from bots to their blocked goals (not rejected by cheaper tests yet). Notice that the just respawned bot on video had enough health to attack first and not retreat, so these lines were not drawn.

**Comment by viciious** [comment]
Failed to see any pink lines in the vid above even after 3 watch-throughs.

Please also post videos showcasing the actual AI behavior from bots' POV. Vid's recorded from human's POV don't provide enough information on the quality of AI's movement or intellectual skills.

Also noticed while testing that the way bots use pg and lg is completely unrealistic (aimbot precision individual shots) and they switch between weapons waay too often, even in close combat. And ye, the way they move is still completely retarded.

**Comment by dnk777** [comment]
> Failed to see any pink lines in the vid above even after 3 watch-throughs.

These lines may be seen in the second part of the video, they are drawn from bot origin to 50hp each time when bot evaluates goals and finds that he has to retreat to some goal and path to the 50hp goal is blocked by enemy

> the way bots use pg and lg is completely unrealistic (aimbot precision individual shots)

Its tweakable, its very minor issue

> they switch between weapons waay too often

It is indended, when weapon selection code is crafted well enough they should select appropriate weapon according to constantly changing combat disposition. 

> And ye, the way they move is still completely retarded.

Did you check the last commit?

**Comment by viciious** [comment]
Now I have. Indeed things have gotten somewhat better now, although bots's aim still wobbles too much and they still love spinning and circling around, as well as aiming straight up. wdm1 is still a very problematic map for them.

Also noticed that due to bot movement circumventing regular player movement, they make no sounds whatsoever.

**Comment by dnk777** [comment]
> Also noticed that due to bot movement circumventing regular player movement, they make no sounds whatsoever.

I change `self->velocity` only in air, and the bot as any player moves silently in air anyway, all dash/jump sounds should not be skipped. All jumps/dashes/ground movements are made as any other player do it.

**Comment by viciious** [comment]
Could have been a glitch but I can swear I couldn't hear any bot sounds at all.

On the movement side, the few remaning major problems are:
- bots bumping into walls a lot
- bots walking past goal items (wdm1 ra is a good example)
- occasional poor navigation even in good conditions: for example, once bot got stuck in ga room on wdm1
- support for teleporters, doors, jumppads and platforms

**Comment by dnk777** [comment]
> bots bumping into walls a lot

They do not know how to walljump yet and just try to avoid walls testing by tracing in runtime, and aircontrol that is not obviously cheating is too weak in this case :) Should be mitigated when walljumping is implemented.

> bots walking past goal items (wdm1 ra is a good example)

Frankly THIS item may be problematic due to weird AAS reachabilities. It may sound discouraging but mapper support is still needed, some junk areas that possibly break path should be marked as «do not enter». Lots of junk areas behind wdm9 pipes near tele-UH are an example. Can't do anything with it.

> occasional poor navigation even in good conditions: for example, once bot got stuck in ga room on wdm1.

Please describe where exactly it has occurred, I have to know is it possible too sensitive runtime obstacles check or just reachabilities glitch.

> support for doors, jumppad and platforms

Jumppad movement should not have very much issues. I have visualized this version of jumppad movement in https://github.com/dnk777/qfusion/blob/dnk777_rebased/source/game/ai/bot_movement.cpp#L312 by drawing a line from a bot to a chosen area center and did not found any glitches while testing.

Platforms are handled there https://github.com/dnk777/qfusion/blob/dnk777_rebased/source/game/ai/bot_movement.cpp#L449, but it may not work as expected. Look at the video above I have posted showing wdm1 reachability mess. There are lots of junk areas on the YA platform that are involved in pathfinding, and when the platform enters TOP state, he possibly tries to follow there areas.

Door support usually depends of AAS pathfinder. On wdm7 bots never chose an obvious path to quad via door, and run around the lower level to reach it. On some other maps, e.g. wbomb2 doors are handled as non-solid things. Door buttons handling is not implemented (and was not in old code).

I agree that movement on wdm1 is problematic, they behave better on 7/9.

**Comment by dnk777** [comment]
On frequent weapon changing, I took care about preventing weapon choice «jitter» those days (bots were changing weapons all the time and not shooting due to weights that are close and changing on each weapon choice reevaluation).
https://github.com/dnk777/qfusion/blob/dnk777_rebased/source/game/ai/bot_brain.cpp#L1135

**Comment by viciious** [comment]
> Please describe where exactly it has occurred, I have to know is it possible too sensitive runtime obstacles check or just reachabilities glitch.

A bot spawned in the ga room on wdm1, decided it'd go for the mega and then started circling around the room: riding on the platform and then dropping back.

> Jumppad movement should not have very much issues.

wdm1 jumppads do cause troubles for bots, especially the one that leads to the ra and eb ammo.

> frankly THIS item may be problematic due to weird AAS reachabilities.

Dunno.. Bots frequently drop off the ra platform by turning 90 degrees away from the goal and walking or jumping off the ledge. I very much doubt AAS reachibilities promote such paths so close to goal items.

**Comment by viciious** [comment]
On the subject of jumppads, try wda5: bots fail to climb the longer jumppad 9/10 times and also tend to look straight up while in the air.

**Comment by viciious** [comment]
> It is indended, when weapon selection code is crafted well enough they should select appropriate weapon according to constantly changing combat disposition. 

Well, as a matter of fact, they should not. In real combat there's simply no time to switch between your weapons, deferring the shot. In real duel combat there's only like two weapons: the primary weapon and the secondary finishing weapon. None is based on the distance, rather more the amount of ammo you have and your estimate of opponents health and armor.

At the moment, bots consider constantly switching their weapons more important than actually shooting them.

**Comment by dnk777** [comment]
I discovered quickly those days why bots fall in places like wdm1 RA (RA item «magnet» starts working prematurely when an item coarse radius (~60-90 units) is reached without checking a ground below a bot, it should start only when a bot is in item AAS area or AAS reachability chain contains only ones of movement type TRAVEL_WALK), but had not enough time to implement and check it to avoid claiming the thing as fixed. The development will be continued soon in 1-2 weeks.

**Comment by dnk777** [comment]
> None is based on the distance, rather more the amount of ammo you have and your estimate of opponents health and armor.

Weapon selection is not based on (only) distance, but on many things. A sophisticated weapon selection code is one of the first things I crafted started working on the project.

> At the moment, bots consider constantly switching their weapons more important than actually shooting them.

Its tweakable by modifying a threshold of difference between suggested and current weapon weights.
Also, I have added a hack (increased threshold) for continuous fire weapons.
https://github.com/dnk777/qfusion/commit/9504d786512c5f20054523166684844aed318769#diff-c0beaa16909d7ce9bec5826781605a25R1362

> it should start only when a bot is in item AAS area or AAS reachability chain contains only ones of movement type TRAVEL_WALK

Fixed in https://github.com/dnk777/qfusion/commit/9b3c42ed45c11b605c8678777b3169e66121d3ac, its unlikely a bot will fail picking up an item like wdm1-RA.

On other movement glitches… I was in hope that AAS will provide a reliable way to handle bot movement, but the code entirely consists of dirty hacks that sometimes fail. I think if bot has troubles in some areas, a mapper should forbid these areas by marking parts of a map as «do not enter» as said in bspc readme. 

What the map compiler produces is a junk, nowhere close to a sane map topology (rooms and hallways) what a human can see and I expected to introduce… It is partially caused by a complexity of the game collision surfaces compared to Q3 ones. These tiny areas that a bot has to navigate through and that may be passed in a single frame only confuse a bot. To mitigate this some kind of weighted interpolation is used by me, and it sometimes may fail. On these videos areas in radius of 48 units around a bot are shown randomly sometimes (to prevent beam events flooding), often it is a complete mess. Note that areas are not boxes but convex objects, their bounding boxes are shown.

http://webmshare.com/YxnA5
http://webmshare.com/xgnZg

**Comment by viciious** [comment]
Have you tried compiling .aas files with the -optimize setting?

**Comment by dnk777** [comment]
An `-optimize` option does nothing but strips AAS file leaving only area AABBs and reachabilities, and that should not be done (AAS geometry such as faces and planes are used, e.g., for target environment sampling for RL aiming providing better quality and/or performance than tracing). AAS area merging is always performed (as it may be seen in logs), but it does not help much as we can see.

https://github.com/TTimo/bspc/blob/master/bspc.c#L819
https://github.com/TTimo/bspc/blob/master/deps/botlib/be_aas_optimize.c#L26

**Comment by dnk777** [comment]
I have finally started a work on the teamplay logic. This video shows squads clustering and reevaluation.

http://webmshare.com/VzV8Z

Lines are drawn from time to time for each pair of squad members (a squad consists of 2-3 bots that should roam and fight together). On 0:15 links reevaluation is clearly visible. They don't try to stick and do actions together, but already share enemies and notify each other. Common squad behaviors such as «assist item pickup», «assist a carrier», «prevent item pickup», «plant an item» will be implemented and made available for GT scripting. Weapon dropping will be added too.

Also, that's a funny video that shows forced backward movement for a bot.

http://webmshare.com/D3eKQ

Obviously a last bot in a squad should move this way to cover back sector when a squad moves together.

**Comment by dnk777** [comment]
A basic weapon dropping for squad members has been implemented. It of course needs more tweaking. On the first video PE#1 grabbed items dropped by itself (because a suppliant YZ2112 preferred to fight with someone).  Weapon drops may be noticed by a blue line drawn between a supplier and a suppliant and a team chat message.

http://webmshare.com/LX77L
http://webmshare.com/vEMMD

Also, basic walljumping has been introduced first time even in my branch.

**Comment by viciious** [comment]
Tbh, bots should only be able to play bomb and duel decently and that's it :P

**Comment by dnk777** [comment]
Thats not a problem to find a duel/bomb game with human players in the current state of the game,
and bots are too stupid for duels anyway. Team-based GTs with items are mine main target,
considering how unpopular are they and how carefully crafted bots can aid them by filling empty player slots on public servers.

**Comment by dnk777** [comment]
I have coded a support of objective-based gametypes (that has defence/offence spots and carriers), 
bots in a team receive defend/attack orders dynamically based on an alert level on a spot (bots report to a team when a defended spot is being attacked, and fair bot vision is used, so it is possible to sneak to a defended spot). This gives a foundation for CTF, CTF:T, Bomb gametypes support. GT scripts modification is required because script interface has been changed, so I have forked scripts too.

For now the CTF script is completed, I even tested that on public with bots and real players.

Also I had to significantly modify AAS routing code to support proper obstacles (areas blocked by enemy) handling. Obstacles were handled aposteriori (a path used to be retrieved and than checked for blocked areas), and that lead to poor bot behaviour because a different non-blocked path might exist. This requires having a separate route cache for each bot, because blocked areas status differs for each bot. Because AAS code had to be modified anyway, I just ripped all useful code from former aas subdirectory, its about 2KLoc and removed other code unrelated to a bot runtime. Introducing a separate route cache is absolutely required for proper obstacles/blocked areas handling, even if it hurts performance a lot by far than everything else (when an area status is changed, a routing cache must be almost entirely dropped). In order to mitigate performance issues I optimized basic AAS routing algorithm too by using a (sorted) heap and not just a linked list of area fringe in breath first search, so it helps to reduce count of iterations 5-6 times.

The thing is almost complete (this means there are no more major native code changes planned), but tweaking, testing and bugfixing will take around 1-1.5 month. 

Also some minor things should be introduced, such as dropping health/armors for teammates in CTF:T (it may be based on already existing and working code of weapon dropping), adding botroam entities support for item-less gametypes, etc.

For the movement issues, fixing the single one opens a can of worms, so the solution is to handle this per map level via AS api and a mapscript (set limitations of movement capabilities for some areas). This will be very cheap (checking current area custom flags), so it wont hurt performance.

**Comment by viciious** [comment]
> For the movement issues, fixing the single one opens a can of worms, so the solution is to handle this per map level via AS api and a mapscript (set limitations of movement capabilities for some areas). This will be very cheap (checking current area custom flags), so it wont hurt performance.

I don't think that expecting mappers to write AS code for bots is a feasible idea. In my 15 years of working on mods and games I've seen only a few artists who were able to cope with code. And by saying they "coped" with it, I'm being overly generous as most of it was a blind copy&paste hackjob.

Neither I intend to provide backwards compatibility for third-party map scripts..

**Comment by dnk777** [comment]
I do not mean _coding_ a logic, just declaring movement limitations for some points on maps. A knowledge of a problematic point origin is all that a mapper need.

Something like this in mapscript (not coded yet): 

`AI::SetAiMovementFlags(Vec3(-1928.0, -213f, -324f), 48.0f, AI_MOVE_ALL_CAPS & ~AI_MOVE_CAPS_DASH);`

On this call all areas in radius of 48 units should be found via `AiAasWorld::BBoxAreas()` and marked as "no-dash". Since a mapscript execution is done once for a map at its loading it wont hurt performance (movement caps are intended to be retrieved by a single array lookup by an area index).

This may be handled on mapping tools level, but
1) It requires modifying of the bsp compiler and the editor.
2) It requires re-compiling a map.

There are usually not more than 2-3 areas on an ordinary map that lead to bot failure. Bot navigation always have been a mapper concern too, and I already tried to minify amount of a mapper job.

**Comment by viciious** [comment]
> 1) It requires modifying of the bsp compiler and the editor.

donotenter and botclip flags are still supported both by the compiler and the engine.

> 2) It requires re-compiling a map.

The opposite also holds true. Modifying brush geometry may require editing the code.

All in all, I strongly suggest you should drop this idea. I never personally liked mapscripts in general. Trust me, possibly using them for bot navigation is bad, _bad_ idea. Fixing poor bot movement with map-specific AS hacks is the road to hell.

**Comment by viciious** [comment]
Oh, btw, in case you haven't noticed, Warsow development has been discontinued.

**Comment by dnk777** [comment]
I has an obvious personal interest to continue it by myself.

**Comment by viciious** [comment]
Btw, have you run performance profiles on your code or tracked its memory footprint/leaks in valgrind?

**Comment by dnk777** [comment]
I ran AddressSanitizer sometimes (it did not spot any leaks), and used valgrind(cachegrind) for profiling.

As I said, toggling areas enabled/disabled status hurts performance a lot but there are no other reliable ways to avoid blocked areas. When an area disabled status has been changed, almost entire routing cache must be cleaned up. In order to mitigate it a bit I sped up (around 1.5x from my measurements) the routing algorithm itself (drained caches should be populated faster). It used a deque of linked `aasroutingupdate_t`, and I use a thing similar to priority queue. On `wbomb5` in a CTF game main loop of `UpdateAreaRoutingCache` function referred below often iterated 17000-18000 times, and after change it has never exceeded a limit of 5000 iterations. Maintaining a heap has some additional cost, thats why resulting performance gain is not as large as it was expected. I also used a optimized for access patterns memory layout for `oldAndCurrAreaDisabledStatus`, inlined `ClusterAreaNum` manually in the loop and precached all references to global objects as stack-allocated locals to avoid pointer-chasing.
https://github.com/dnk777/qfusion/commit/693232e0204196ad638a6eabe1beff4fdc0aeabc

Also I took care about reducing count of routing calls during a search for a goal by selecting not more than 16 best items and testing routes only then in hope the router won't fail for at least a single item.
https://github.com/dnk777/qfusion/commit/26e0ff1c92ad9e378da6c6ca00c2b9c24e668c1e

On a listen server on huge maps performance drops are very noticeable.
On a dedicated server it should be satisfiable, I have never seen drops below 60 fps on very huge maps in CTF/CTF:T (these gametypes use most complicated bot logic).

Performance have always been an thing I took care about during the development of this code, I optimized features once they were well established many times. The most common optimization pattern I used is to reduce number of expensive trace/routing calls by trying to avoid these calls by doing all cheap rejection tests first, pre-selecting various items lists by filtering and limiting inputs and calling expensive calls only then for selected items instead of doing the calls mentioned above in a loop on a raw list. There are numerous examples of this approach in the code, the most recent is https://github.com/dnk777/qfusion/commit/2af83f62cd5ca5d81a8a4626c86378acdc429372

**Comment by viciious** [comment]
> below 60 fps

How huge was the drop compared to stock Warsow? The minimum acceptable framerate for Warsow is generally considered to be 125fps.

**Comment by dnk777** [comment]
The drop is entirely caused by code of my bots, and I have mentioned above the main source of performance problems. The performance impact after introduction of a separate route cache for each bot that has an ability to dynamically enable/disable routing areas was noticeable, but bot behaviour benefits a lot from this change, and that is clearly visible. Without proper exclusion of blocked areas from routing bots used to either rush headlessly in a crowd or be blocked in a corner. The thing is in final stage but still is in active development, so I will try to optimize it more. I do not try to advocate myself but any arena game AI that pretend to be «done right» is known as a CPU burner (I am speaking about UT and especially Q3 CMPA bots). On a dedicated server I had a solid gameplay experience, even tested that in public. For low-end clients like Android a separate build that have some CPU-expensive features cut off may be used. Gameplay on Android is different from the vanilla game anyway, and hosting a public game on a phone is a joke, so barely anyone will notice any difference.

**Comment by viciious** [comment]
Have you considered limiting routing cache resets to only 1 bot per frame? Pretty sure that will not cause any noticable impact on the gameplay. If that doesn't work out, the only other remaining option is be multi-threading, which is something modern games do for the AI anyway. But that would take significant effort.

**Comment by dnk777** [comment]
Routing cache may be reset not more frequently than 4 frames for `sv_fps="60"`.
I took care about uniform distribution of workload from the very beginning by splitting all updates for all ai-related classes in `Update()` into `Frame()` (that contains urgent updates that cannot be skipped) and `Think()`, where `Think()` is scheduled by some global manager. https://github.com/dnk777/qfusion/commit/7d7aba8f9c760d61b5cdaa68cac914b4e7ad3e14
For example, in ffa game with 8 bots `Think()` should be called on each frame for 2 bots.
In ctf game workload for 8 bots is distributed this way:

frame N+0: Alpha team brain thinks (it may be as expensive as thinking of four bots)
frame N+1: Four Alpha bots think
frame N+2: Beta team brain think
frame N+3: Four Beta bots thinlk

What hurts performance itself, is not the cache resetting itself but re-populating drained cache in following routing calls, so AAS routing should be sped up more than in my attempt https://github.com/dnk777/qfusion/commit/693232e0204196ad638a6eabe1beff4fdc0aeabc.

For a dedicated server multithreading is not an option (because usually even multiple instances of a server are run on the same machine). But dedicated server fps is limited to 60 anyway. I think the common good VPS hardware (2+ GHz) is capable of running 2 CTF games on the same machine (when I tested a server on an unknown VPS it shown 40-70% CPU load).

For a listen server this may be an option, since client updates frequency is vital. `AiAasRouteCache` ops are not thread-safe, but bots use separate route caches now (because set of blocked areas is different), so updates of routing cache for more than 1 bot that should occur in a single frame may run in parallel in separate threads. All ops that heavy rely on routing are easy to isolate, so an ordinary fork-join pattern may be used.

**Comment by viciious** [comment]
> What hurts performance itself, is not the cache resetting itself but re-populating drained cache

Obviously, but it's the cache drain that causes the re-populating. 

> All ops that heavy rely on routing are easy to isolate, so an ordinary fork-join pattern may be used.

No, it may be not. Join is going to cause the joining thread to block until the joined thread exits, which is pretty much the same thing as running the job in a single thread. Instead, the worker threads should be assigned jobs, which they would perform in the background and notify the scheduling thread upon completion.

I'm very reluctant to provide the MT functionality in the game module though. Having the AAS routines alone there sucks badly enough already.

> For a dedicated server multithreading is not an option (because usually even multiple instances of a server are run on the same machine).

You're confusing physical CPU's with virtual CPU cores/hyper-threading pipelines. Instancing game servers inside docker-alike containers with a variable number of CPU cores is pretty much the de-facto modern day standard.

**Comment by dnk777** [comment]
Looks like a false alarm for me. Most drops were caused by my client, looks like an occasionally measured debug build or driver issues. I thought you have tested the build and experienced same drops as me, haven't you? On regular maps client fps usually never drops below 200 fps, and even on outstanding by navigation complexity maps wbomb5, wctf2 where I witnessed a bad performance it never goes below 120.

I forgot to say, a routing cache is never cleaned neither each frame, neither each `Think()`. Updating disabled areas status is tied to combat task selection or update which should be never called more frequently than 200-300 milliseconds.

**Comment by viciious** [comment]
I tested the branch a couple of months ago. I did experience noticable fps drops on wdm1 even back then so I assumed things have only gotten worse since then.

**Comment by viciious** [comment]
Note that the longer the branch goes on, the fewer chances it has to ever get merged. There's no physical way I can possibly review 200+ commits with tens of thousands of changed code lines. Not to mention the fact that you seemingly prefer to ignore my comments.

**Comment by dnk777** [comment]
It is almost complete, I will commit the last major native code change in a few hours. But there are known minor bugs, so the entire next month the branch will be actively tested for stability, including public tests and some minor changes to conform your requirements will be made.

From the script part I did not touch the bomb script yet (it allows this gametype to run but bots have poor behaviour). The last changes were almost entirely related to support of most CTF:T features from AI side, and players highly rated it. Adding a good bomb support does not require changes in native code (since I have coded a support for generic gametypes with defence/offence spots).

**Comment by viciious** [comment]
hi @dnk777 

Any estimate on when you believe you're done with the changes? Not that I'm particularly fond of some of them (the introduction of "script weapons" is one of them) but I really need to do some filter-branching on master and would like to have this merged ASAP.

**Comment by dnk777** [comment]
That branch is frozen, no changes are planned. This means all gametypes have basic support and there should not be any critical bugs (I ran the code under AdrressSantizier and fixed detected bugs). So should be ready for merging.

This however does not mean I consider AI development complete and will abandon the development complete.  They give a good illusion of a real match in CTF, CTF:Tactics (thats why I put effort in CTF:T, playing it was funny) but in bomb individual bot weakness is noticeable and very distracting, especially when bot is on his own at the end of round and you have to spectate it.  Thats because I did not put much effort in individual bot AI, most of the time was spend on fixing movement, teamplay logic (which is quite solid so will be kept almost as is) and supporting gametype features via scripts. 

TL, DR:

1) Individual bot AI should use GOAP as described in 
http://web.media.mit.edu/~jorkin/gdc2006_orkin_jeff_fear.pdf. 
The code now as is is a messy FSM where state transitions are spread over it, so modification and tweaking it is a pain, and thus forcing bots to expected behaviour is a quest (for example bots are too easily distracted by enemies in bomb and ignore planting often). Thats how planning code will look like https://gist.github.com/dnk777/f2aa5bd9312bfb101b3b49ab6993121d (the code is in my private sandbox).

2) Old navigation nodes will be reused as anchors for tactical spots for proper positioning in fight, properties of nodes as tactical spots and mutual spots reachability/visibility/other relationship will be precomputed at map loading.

3) These changes will be implemented but 2-3 months later, I have already spent 10x more time than expected and want to take a rest of it.

**Comment by viciious** [comment]
Well, since there have been no pull requests from you, I'm assuming you would rather prefer to keep your code out of the mainline and maintain it yourself

**Comment by viciious** [comment]
Also note that now that  no longer maintain Warsow, I'm probably going to eventually deprecate all gametypes except for duel, bomb and rekt (which remains unpublished).

**Comment by dnk777** [comment]
1) Scripts are modified, so if you need rekt gametype support give the script to me and I will modify it to be able to launch it.
2) I admit that mine code has poor quality and there might be several minor glitches. Thats because I was completely unaware both of AI domain and the engine starting the project. I got a clear vision of what AI needs only when I crafted support for all gametypes. I will address that in a patch later along with a proper support for bot planning, positioning and reasoning via GOAP (which I have already started to implement in a private repo as posted in the gist above; the code base is almost entirely kept but is much cleaner and loosely coupled from the first look than existing). 
3) The code and script interface will be documented in the project wiki when I submit the mentioned patch later.

**Comment by dnk777** [comment]
I ought to write some kind of report on bot development, since it might seem abandoned, but it is not.

I feel a bit sorry for that I has to diverge my branch a lot (again), but I have an excuse, the introduced features took/are taking significant amout of time to implement and test to ensure they work as indended before I am able to claim they are completed, and they are (sometimes mutually) dependend, so I cannot split feature sets and work separately.

First, the GOAP (Goal Oriented Action Planning) was introduced as a method of implementing bot decision making in well-structured way instead of if-conditions and state primitive variables spread over the code. The GOAP is generally described in the paper above. However mine implementation is very different because they use search from the goal world state to the initial one, and I do the opposite (this makes using procedural conditions natural, and also makes the suggested action sequence way more controllable). Tbh GOAP is a pretty trivial thing once you get it. Also a script interface to the bot planner is added. There is an ability to define goals and actions (sequence of actions satisfy goal) in scripts, and also add script-defined actions to an existing native goal. For example if a player class has a cloaking ability, you can define a `SneakCloakedToNavTarget` action and tie it to the native `PickupItemAction`, and then bots would use it to sneak into a flagroom cloaked, or just sneak to an item/tactical spot to prevent being spotted and damaged, and it does not require hardcoding custom gametype features in native code.

Then I had to add an evolution algorithms support for action/goal weights. First tests of GOAP results often shown awkward results. There were lots of numeric constants in the code that affect bot behaviour, modifying each of these constants might change things a lot, and manual tweaking is just counter-productive when a substantial count of these constants is reached. All constants that tied to GOAP goals/actions are converted to config vars and hierarchical groups of such vars, and are serialized to a text format. A basic support for randomizing config weights for bots at start based on a previous best weight config and selecting new best weight config based on a match results is added. There is an ability to override randomization/selection of best config rules in scripts, and add vars and var groups in scripts for script-defined actions that are serialized along with the native part of weight config. Separated weight configs for each gametype and even each map are supported via native code. There is also a possibility to fine-grain control bot weight configs in scripts, for example it is possible to assign class-specific weight config to the bot when its player class gets changed.

Bot evolution feature has proven to be working, there were public playtests in January that included 40 (!) bots playing on a single map (wdm7) in ffa mode, and is was noticeable that best bots in selection start play much less awkward after 2-3 days of evolution. IIRC Q3A bots use a similar feature (not for final releases but in development to get a good weight config).

Frankly even after implementation of these two features results were not satisfiable for me. Still present bot movement troubles were hiding improvement in decision making. Then i noticed that Q3 bots predict movement results for several frames ahead in runtime. I known the existence of such functionality in AAS code base but thought it is used only in bspc. From what i read UT games predict bot movement ahead too up to several dozens of frames ahead. I have however introduced a much more advanced system to fix the bot movement ffs.

In Q3A, a movement action gets selected for the current game frame, ucmd/angles are produced and then a movement gets predicted for several frames ahead with a fixed input defined on the first frame, and results of the prediction are checked in the current game frame. This gets repeated on next game frame. I modify ucmd/angles according to the prediction frame state on each frame selecting a suitable movement action, use variable timestep and caches of helper tests for each frame. If a movement action sequence application leads to a failure in the predicted state, the prediction stack can be rolled back, and action can be restarted with different params or disabled for further planning. Prediction results for frames are saved and might be used in next game frames until a misprediction error accumulates, and this allows to mitigate possible performance issues from this approach.

The movement prediction feature is still in development, tweaking and testing and cannot be considered ready for public testing but already produces solid results. Bots (almost) never miss item pickups, bump into walls unintentionally and go into undesired areas, all of this mostly keeping bunny-hopping. I would stay with basic movement for this branch to avoid it further growth (for example current bots in master aggressively use rocketjump shortcuts, and I have rocketjump movement prediction code that simulates knockback, but I keep this code private without intentions to include in the branch because it would take lots of time to integrate and test it).

I am going to add only critical or very, very important fixes in my branch (I can suggest lots of improvements but they are to be submitted later in separate branches), but I have to spend a month of tweaking/testing the code, especially for all gametypes, to ensure it works.

**Comment by viciious** [comment]
I've been following your progress on the GOAP branch, nice to see you're still working on it. Although I can see how troublesome the eventual merge is probably going to be.

It'd be nice if the new AI code was reusable for monster intelligence, not just bots alone.

**Comment by dnk777** [comment]
Poor movement has been the bane of the AI since I have started working on bots. Luckily there is a solution that have started producing solid results, the bot movement prediction system. This system allows to test all applicable movement actions in a sandbox and save the path of a successfully applied action for reusing in future frames. Its so powerful so I had a surprise when it rejected almost all bunnying attempts produced by a code that was similar to the existing code in the master branch. Writing movement actions that pass all prediction tests is not an easy task and requires lots of tweaking and ideas testing. Each commit related to the movement system since March has been updated and rebased not less than 20-30 times (and still might be). I tried to avoid adding extra commits in the branch but I had to do it for some important things (and also I rather add a fix commit now than update and rebase an existing one that is dated 1-2 month ago, I want to avoid rebasing a long commits chain due to it being quite error-prone).

Here is the video of bot movement prediction system in the current state. A path built for an action that can be applied to the current movement state is shown. The first part of the demo was recorded using notarget cheat, and also timescale is a bit lower than the normal.

http://webmshare.com/VJNOj

I did not try to hide any glitches that appear. There is still a known bug very noticeable on the video. A misprediction occurs way too often and the entire predicted path gets invalidated leading to losing speed/switching to a dummy movement, and the bot movement is rather poor due to it. Ideally bot movement prediction system should not get activated again until the bot origin reaches the end of a predicted path. I need to investigate a source of the bug since I have tried to visualize the origin delta error and the error pattern looks regular, like a position with  +/-16 millis offset of an expected origin. Its probably a some issue related to predicted action timestamps. Also switching to a dummy, non-predicted movement is noticeable in the beginning of the video near the A-spot UH (no path is shown), and it leads to weird movement especially in the mentioned area. This kind of movement should be avoided, but I still did not manage to write actions that are applicable (pass all prediction tests) for all situations.

There are already great results in such areas as:

1) Crouchsliding, its noticeable at 0:41 on the video, its very cool to see bots doing it.
2) The combat movement that is purely amazing, they have mad dodging, and also notice it is predicted (as almost any other kind of movement) and should never lead to falling into pits/lava/losing a tactical position. I hope other movement actions would have the same quality as the combat one.

My plans consist of fixing the mentioned misprediction bug, adding some minor features (like forcing backward movement switching from combat to running to always face an enemy) that totally are estimated in a minor amount of code (about 200-250 loc totally) and testing for critical bugs and gametype compatibility. I still need an extra month to finish it.

**Comment by dnk777** [comment]
> It'd be nice if the new AI code was reusable for monster intelligence, not just bots alone.

I have no idea how monster code would look like if somebody decides to implement one, but all 3 major features introduced in the branch (GOAP / genetic weigths optimizer / movement prediction system) are very generalizable with relatively low effort (and other bot code like aiming, squad formation, etc too). For example, arbitrary script attachments to the GOAP logical world state, scripted GOAP goals and actions are already allowed.

On merging... rebasing it over master is very painful (mainly due to uncrustify applied right after the branches have diverged), but merge with master without rebasing is pretty straightforward (especially if I run uncrustify as a last commit in my branch).

**Comment by dnk777** [comment]
Here is the list of issues I plan to continue working on and sumbit series of tiny patches for each one:

June:
* Switch back to an automatic tactical spots detection based only on AAS and heuristics making nav files obsolete again.
* Conform to the crouchsliding being turned on/off in the master by pmove flags. I plan to keep skimming for bots since it helps in some situations  and is barely noticeable from a 3rd person POV (it is not a good idea to spectate a bot anyway because ones move hackish not to mention the absence of brain causing nausea for a viewer).
* Add PMove() flags to skip some expensive collision computations when PMove() is used for bot movement prediction and these computations are redundant for a given context.

July:
* Only basic GOAP actions set was added. Even melee attacks are missing (because they should be planned - (get close to the enemy by some way without being significantly hit -> attack melee) and there are no GOAP actions yet that produce satisfiable GOAP world state). Also, existing GOAP actions should be checked and tweaked.
* Compatibility with 64-bit timers has not been tested well, just the bot behaviour seems the same and there are no type warnings.
* Tweak the existing bot movement, I have a hope that way better movement quality could be jammed from existing movement actions.
* Check compatibility with all currently present gametypes and manage to get a good bot behaviour for each gametype.
* Discover the source of bot movement misprediction that leads to way too frequent movement path invalidation causing increased CPU load.

Later:
* Re-add the excessive weaponjumping present in the current master and temporarily removed when the bot movement prediction system has been introduced.
* Optimize AAS code/memory layout, decouple file and in-memory AAS representations.
* Enemy/weapon/item selection code is dated by the beginning of the advanced bots development and should be improved.
* Add GOAP planning for AI squads too (goals like: stay together, run together to a nav target, attack a spot coordinately, defend a carrier and set of corresponding actions).
* Allow mixed (bot/human) squads (bots currently drop weapons only to other team bots).
* Add bot commands and chat listening (if a human cries "AAAAA" in team chat while playing a bomb match, bots should response to it).

**Comment by dnk777** [comment]
> Add PMove() flags to skip some expensive collision computations when PMove() is used for bot movement prediction and these computations are redundant for a given context.

To explain this a bit: on each movement prediction step a coarse environment test might be performed. If a predicted origin is inside bounds of a huge AAS area, its very cheap. In worst case, a single box trace test is performed. If this coarse test succeeds, there is no need for further obstacle checking. Results are shown on this video http://webmshare.com/m8OOG. If this coarse test has not detected any obstacles (green lines drawn from tested box mins to maxs)  and the bot is above ground, PMove() should skip very expensive "stepmove/slidemove" collision computations. I has not added PMove() hint flags in this branch trying to keep all introduced changes inside the `/ai` subdirectory. If these hints are added, they allow to significantly reduce CPU load or/and allow better quality of bot movement action checks.

Here are some issues I forgot to add in the list above:

* Sometimes a bot might be blocked by environment or by some other reason. Old code for escaping from blocked state has been temporarily removed. Escaping from blocked state should be a GOAP goal the planner is aware of to prevent clash with other GOAP goals/actions execution.

* Bots are deaf in the current state making them easy targets. They should react to sound events (some hooks should be added in the `game` module and a sound listener/detector should be added from AI side), and also deduce an information about enemy from seeing enemy shots/projectiles.

* Spamming to potential enemy positions is missing. It can be added easily as an action that satisfies already present `ReactToEnemyLost` GOAP goal.

**Comment by dnk777** [comment]
Also I have rewritten the first commit that introduces the bot movement prediction system to skip all collision tests against entities in PMove() call (the code commits are self-explanatory). I have tested 32 bots playing FFA on a listen server run on 3 GHz Intel and got these results: 130-170 FPS on wbomb1, 150-200 FPS on wdm8. (Note: I have used the mine proposed CM improvements). If PMove() hints are added later, there should be a significant performance boost too.

```
@@ -384,6 +384,12 @@ static void Intercepted_PMoveTouchTriggers(pmove_t *pm, vec3_t previous_origin)
     game.edicts[pm->playerState->playerNum + 1].ai->botRef->OnInterceptedPMoveTouchTriggers(pm, previous_origin);
 }
 
+static void Intercepted_Trace( trace_t *t, vec3_t start, vec3_t mins, vec3_t maxs, vec3_t end,
+                               int ignore, int contentmask, int timeDelta )
+{
+    trap_CM_TransformedBoxTrace(t, start, end, mins, maxs, nullptr, contentmask, nullptr, nullptr);
+}
+
 void BotMovementPredictionContext::OnInterceptedPredictedEvent(int ev, int parm)
 {
     switch (ev)
@@ -1023,8 +1029,21 @@ void BotMovementPredictionContext::NextMovementStep()
 
     this->frameEvents.Clear();
 
+    // We currently test collisions only against a solid world on each movement step and the corresponding PMove() call.
+    // Touching trigger entities is handled by Intercepted_PMoveTouchTriggers(), also we use AAS sampling for it.
+    // Actions that involve touching trigger entities currently are never predicted ahead.
+    // If an action really needs to test against entities, a corresponding prediction step flag
+    // should be added and this interception of the module_Trace() should be skipped if the flag is set.
+
+    // Save the G_GS_Trace() pointer
+    auto oldModuleTrace = module_Trace;
+    module_Trace = Intercepted_Trace;
+
     Pmove(&pm);
 
+    // Restore the G_GS_Trace() pointer
+    module_Trace = oldModuleTrace;
+
     // Update the entity physics state that is going to be used in the next prediction frame
     entityPhysicsState->UpdateFromPMove(&pm);
     // Update the entire movement state that is going to be used in the next prediction frame
```

**Comment by dnk777** [comment]
The following contents https://gist.github.com/dnk777/063785d84087a446623ce69bbfbd5087 is expected to be in `/ai` game filesystem subdirectory under the `generic.weights` name. It is a result of 70+ hours of bot evolution in FFA mode. Otherwise the currently hardcoded bot behaviour might seem weird (and really is).

These weights are of course not final, and would be threated differently when I tweak the bot logic. For example, there is a separate `ReactToThreatGoal` (it gets activated when a bot gets a damage from an attacker). I have added a concept of "points keep in fov" as one of the last commits in this branch. This point is set to an enemy origin/just lost out of sight enemy origin. A bot might invert his movement (move backwards) trying to keep this point in fov if the point is present and it is necessary. I have an idea of setting this point to a possible threat source, and thus weights of the separate `ReactToThreat` goal should be lower (but this goal is still needed in some cases).

Also weights for objective-based team gametypes should be assigned dynamically for defenders/attackers/different player classes when a bot changes its role/class. It is also remains a thing to be done in future.

**Comment by viciious** [comment]
> The following contents https://gist.github.com/dnk777/063785d84087a446623ce69bbfbd5087 is expected to be in /ai game filesystem subdirectory under the generic.weights name. It is a result of 70+ hours of bot evolution in FFA mode. Otherwise the currently hardcoded bot behaviour might seem weird (and really is).

Please submit a pull request here: https://github.com/Qfusion/rapid/

I'm closing this issue as it's become too generic and huge to remain comprehensible. Please restrain yourself to smaller tasks further on.
